hcl2template/types.packer_config.go

```diff
@@ -311,13 +352,92 @@ func (cfg *PackerConfig) evaluateDatasources(skipExecution bool) hcl.Diagnostics
 			})
 			continue
 		}
+
 		ds.value = realValue
 		cfg.Datasources[ref] = ds
 	}
 
+	// Now that most of our data sources have been started and executed, we can
+	// try to execute the ones that depend on other data sources.
+	for ref := range dependencies {
+		_, moreDiags, _ := cfg.recursivelyEvaluateDatasources(ref, dependencies, skipExecution, 0)
+		// Deduplicate diagnostics to prevent recursion messes.
+		cleanedDiags := map[string]*hcl.Diagnostic{}
+		for _, diag := range moreDiags {
+			cleanedDiags[diag.Summary] = diag
+		}
+
+		for _, diag := range cleanedDiags {
+			diags = append(diags, diag)
+		}
+	}
+
 	return diags
 }
 
+func (cfg *PackerConfig) recursivelyEvaluateDatasources(ref DatasourceRef, dependencies map[DatasourceRef][]DatasourceRef, skipExecution bool, depth int) (map[DatasourceRef][]DatasourceRef, hcl.Diagnostics, bool) {
+	var diags hcl.Diagnostics
+	var moreDiags hcl.Diagnostics
+	shouldContinue := true
+
+	if depth > 10 {
+		// Add a comment about recursion.
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Max datasource recursion depth exceeded.",
+			Detail: "An error occured while recursively evaluating data " +
+				"sources. Either your data source depends on more than ten " +
+				"other data sources, or your data sources have a cyclic " +
+				"dependency. Please simplify your config to continue. ",
+		})
+		return dependencies, diags, false
+	}
+
+	ds := cfg.Datasources[ref]
+	// Make sure everything ref depends on has already been evaluated.
+	for _, dep := range dependencies[ref] {
+		if _, ok := dependencies[dep]; ok {
+			depth += 1
+			// If this dependency is not in the map, it means we've already
+			// launched and executed this datasource. Otherwise, it means
+			// we still need to run it. RECURSION TIME!!
+			dependencies, moreDiags, shouldContinue = cfg.recursivelyEvaluateDatasources(dep, dependencies, skipExecution, depth)
+			diags = append(diags, moreDiags...)
+		}
+	}
+	// If we've gotten here, then it means ref doesn't seem to have any further
+	// dependencies we need to evaluate first. Evaluate it, with the cfg's full
+	// data source context.
+	datasource, startDiags := cfg.startDatasource(cfg.parser.PluginConfig.DataSources, ref, true)
```

Comment:
Do you think we should check for errors in `diags` here, before continuing ?

Reply:
Sure! I just added an error check to the earlier recursion loop which makes sure that by the time we get here we've already checked diags.